---
title: "3.18 Managing Genotypic Data in VCF"
layout: doc_page
---

<!-- TOC-START -->
* TOC
{:toc}
<!-- TOC-END -->

### 3.18.1 Uploading Genotyping Data in VCF

Genotyping data in VCF can be loaded from the web-interface. Breedbase can store any genotypic variants from a VCF, allowing for polyploids, structural variants, etc. without problems.

To begin go to Manage->Genotyping Plates and click the button seen below:
Note that you do not need to have genotyping plates uploaded to upload VCF data; you may upload genotyping data to accessions or you can upload genotyping data for tissue samples in genotyping plates.

![]({{"assets/images/manage_genotyping_data_upload_button.png" | relative_url }})

The workflow begins with an intro:

![]({{"assets/images/manage_genotyping_data_upload_dialog_intro.png" | relative_url }})

On the following step in the workflow, a genotyping project is defined or selected. A genotyping project is a high-level entity for grouping several genotyping events. It is defined with a name, description, name, breeding program, and genotyping facility (IGD, Intertek, etc.).

![]({{"assets/images/manage_genotyping_data_upload_dialog_project.png" | relative_url }})

The following step is to define or select a genotyping protocol. A genotyping protocol represents the set of markers being called against a specific reference genome. A genotyping protocol is defined with a name, description, reference genome name, species name, and a location of data generation. Note in the picture that you can select whether the samples in your file are accessions or tissue samples in the database; tissue samples are for when a genotyping plate is stored in the database. There is an option to parse the sample names for appended sequencing numbers from IGD, where the sample names are like "accession:igdnumber".

![]({{"assets/images/manage_genotyping_data_upload_dialog_protocol.png" | relative_url }})

The final step is to select the VCF from your computer and upload it. The web interface can be used to upload files arbitrarily large; it is a NGINX configuration to set this size.

![]({{"assets/images/manage_genotyping_data_upload_dialog_vcf.png" | relative_url }})

### 3.18.2 Searching and Downloading Genotyping Data in VCF Against Phenotyping Data

The Search Wizard is the primary means of querying data in the database. Go to Search->Wizard to begin.

Once genotyping protocols are stored, select Genotyping Protocols from the first dropdown menu. Then if you select one or more and select Accessions from the second dropdown menu, you will see the accessions for which genotypes were stored. As seen in the following picture, there is a section for filtering genotypes by chromosome, start position, and end position. Genotypes can be downloaded in VCF or DosageMatrix formats.

![]({{"assets/images/manage_genotyping_data_wizard_download_prot.png" | relative_url }})

Using the "Default genotyping protocol" which is configured in a system, you can query over field phenotypic evaluations before downloading genotypes and phenotypes.

![]({{"assets/images/manage_genotyping_data_wizard_trial.png" | relative_url }})

