
use strict;
use CXGN::Page;

my $p = CXGN::Page->new();

$p->header;

print <<HTML;

<h1>Maloof Lab, UC Davis, USA</h1>

<img src="/static_content/community/feature/201204/Maloof_Lab.jpg" border="0" />
<font style="font-style:italic">Maloof Lab: (Left to right). Back: Max Mumbach, Dan Fulop, Susan Bush, Christine Palmer, Amanda Schrager, Leonela Carriedo.  Middle: Mei Yi Zhang, Kazunari Nozue, Julin Maloof, Robert Harris, Guo Gui Ning.  Front: Upendra Devissetty, Mike Covington.</font>


<p>Our research is focused on understanding the genetic and molecular basis of plant photomorphogenesis - how plant growth and development change in response to light.  A major part of our research is directed towards understanding the molecular basis for natural variation in light responses between populations and species. </p>

<img src="/static_content/community/feature/201204/TomatoShade.jpg" />
    <p>
<font style="font-style:italic">Tomato Shade: Plants grown in simulated sun (left) or simulated shade (right).  S. chilense shows a stronger response than S. lycopersicum to simulated shade.</font></p>

<p>Because plants depend on light for photosynthesis, they have developed a system to detect competition from their neighbors and to respond accordingly.  Light reflected from neighboring plants is reduced in its red to far-red (R/FR) ratio. The phytochrome (PHY) photoreceptors perceive low R/FR and trigger shade-avoidance responses that allow plants to compete for light.  These responses include increased stem and petiole elongation, accelerated flowering, and reallocation of resources from leaves, fruits, and tubers to stems.  Shade avoidance is of interest to developmental biologists studying how these responses are generated, agricultural engineers interested in increasing crop yield by reducing these responses, and evolutionary biologists interested in this preeminent example of adaptive phenotypic plasticity.</p>

<p>The correct response to a specific light cue depends on the environment. Plants native to sunny environments are sensitive to foliar shade and show robust shade-avoidance. In contrast, shade avoidance is reduced in plants native to constitutively shady environments. Such changes in shade-avoidance response are adaptive and are due to heritable genetic differences. We are interested in determining the genes responsible for adaptive changes in light response, the mechanisms by which changes in these genes affect light signaling, and the evolutionary forces that have acted upon them. </p>

<p>We have found significant variation in photomorphogenesis and shade avoidance in tomato and its wild relatives.  For example S. lycopersicum var. M82 is a mild shade avoider whereas S. pennelli has a strong response.  We are using the S. pennellii X M82 introgression lines to map QTL and clone genes underlying this variation.  To aid in this effort, in collaboration with Neelima Sinha\'s group, we have performed extensive RNAseq of S. pennellii, M82, the introgression lines, and several other wild tomato relatives.  This has allowed us to define hundreds of thousands of SNPs differentiating these species and to study the patterns of nucleotide and gene expression differences.</p>
<p>We also study shade avoidance and quantitative genetics in Arabidopsis thaliana and Brassica rapa.</p>

<img src="sgn_static_content/community/feature/201204/IL4_3.png" border="0" />

<font style="font-style:italic"><p>IL4_3:  RNAseq analysis of the S. pennellii X M82 introgression line IL4_3.  Each segment around the circle represents a chromosome.  From outside to inside the rings show: log fold change of genes differentially expressed between M82 and IL4_3 (the black line represents 0); the genotype of each chromosomal region (blue = M82, green = S. pennellii); genes regulated in trans from the introgression (light green: transgressive expression; dark green: S. pennelli-like expression).</p></font>


<h3>Contact</h3>
Julin Maloof, PhD<br />
Department of Plant Biology<br />
College of Biological Sciences<br />
University of California, Davis<br />
Davis, CA 95616<br />
+1 530 752 8077<br />
<a href="mailto:jnmaloof\@ucdavis.edu">jnmaloof\@ucdavis.edu</a><br />
<a href="http://malooflab.openwetware.org/">http://malooflab.openwetware.org/</a><br />





<h3>Selected Publications</h3>
<ol>
<li>Seymour, D. K., Filiault, D. L., Henry, I. M., Monson-Miller, J., Ravi, M., Pang, A., Comai, L., Chan, S. W. L., and Maloof, J. N. (2012). Rapid creation of Arabidopsis doubled haploid lines for quantitative trait locus mapping. Proc Natl Acad Sci USA 109, 4227-4232.</li>
<li>Filiault, D. L., and Maloof, J. N. (2012). A Genome-Wide Association Study Identifies Variants Underlying the Arabidopsis thaliana Shade Avoidance Response. PLoS Genet 8, e1002589.</li>
<li>Chitwood, D. H., Headland, L. R., Filiault, D. L., Kumar, R., JimÃ©nez-GÃ³mez, J. M., Schrager, A. V., Park, D. S., Peng, J., Sinha, N. R., and Maloof, J. N. (2012). Native Environment Modulates Leaf Size and Response to Simulated Foliar Shade across Wild Tomato Species. PLoS ONE 7, e29570.</li>
<li>Jimenez-Gomez, J. M., Corwin, J. A., Joseph, B., Maloof, J. N., and Kliebenstein, D. J. (2011). Genomic analysis of QTLs and genes altering natural variation in stochastic noise. PLoS Genet 7, e1002295.</li>
<li>Arana, M. V., MarÃ­n-de la Rosa, N., Maloof, J. N., BlÃ¡zquez, M. A., and AlabadÃ­, D. (2011). Circadian oscillation of gibberellin signaling in Arabidopsis. Proc Natl Acad Sci USA 108, 9292-9297.</li>
<li>Chang, C.-S. J., Maloof, J. N., and Wu, S.-H. (2011). COP1-mediated degradation of BBX22/LZF1 optimizes seedling development in Arabidopsis. Plant Physiol 156, 228-239.</li>
<li>Nozue, K., Harmer, S. L., and Maloof, J. N. (2011). Genomic analysis of circadian clock-, light-, and growth-correlated genes reveals PHYTOCHROME-INTERACTING FACTOR5 as a modulator of auxin signaling in Arabidopsis. Plant Physiol 156, 357-372.</li>
<li>Brock, M. T., Dechaine, J. M., Iniguez-Luy, F. L., Maloof, J. N., Stinchcombe, J. R., and Weinig, C. (2010). Floral genetic architecture: an examination of QTL architecture underlying floral (co)variation across environments. Genetics 186, 1451-1465.</li>
<li>Jimenez-Gomez, J. M., Wallace, A. D., and Maloof, J. N. (2010). Network analysis identifies ELF3 as a QTL for the shade avoidance response in Arabidopsis. PLoS Genet 6.</li>
<li>Jimenez-Gomez, J. M., and Maloof, J. N. (2009). Sequence diversity in three tomato species: SNPs, markers, and molecular evolution. BMC Plant Biol. 9, 85.</li>
<li>Filiault, D. L., Wessinger, C. A., Dinneny, J. R., Lutes, J., Borevitz, J. O., Weigel, D., Chory, J., and Maloof, J. N. (2008). Amino acid polymorphisms in Arabidopsis phytochrome B cause differential responses to light. Proc Natl Acad Sci USA 105, 3157-3162.</li>
<li>Nozue, K., Covington, M. F., Duek, P. D., Lorrain, S., Fankhauser, C., Harmer, S. L., and Maloof, J. N. (2007). Rhythmic growth explained by coincidence between internal and external cues. Nature 448, 358-361.</li>
</ol>


HTML

    $p->footer();
