#!/usr/bin/perl

######################################################################
#
#  Program  :  $Id$
#  Author   :  $Author$
#  Date     :  $Date$
#  CHECKOUT TAG:  $Name:  $
#
#  This program is barely a program at all, although it does pull
#  somewhat from the database and so ended up a program.
#
#  Nevertheless, the point here is really to generate a largely 
#  static page, one which serves to explicate the Overgo Plating
#  process.
# 
######################################################################

use strict;
use CXGN::Page;
use CXGN::DB::Connection;
use CXGN::DB::Physical;
use CXGN::Page::FormattingHelpers qw/blue_section_html page_title_html/;

# Presets.
my $overgo_stats_page = 'overgo_stats.pl';
my $physical_map_page = '/cview/map.pl?map_id=9&amp;physical=1';
my $physical_map_overview = '/cview/map.pl?map_id=9&amp;physical=1';
my $soop_home = 'http://genome.nhgri.nih.gov/soop/';
my $agi_page = 'http://www.genome.arizona.edu/fpc/tomato/';
my $bac_search_page = '/search/direct_search.pl?search=BACs';

# Connect to the db.
my $dbh=CXGN::DB::Connection->new('physical');

# Prepare the page.
our $page = CXGN::Page->new( "About the overgo plating process", "Robert Buels");

# Find out how many plates have been processed to date.
my $plates_sth = $dbh->prepare("SELECT COUNT(DISTINCT plate_number) FROM overgo_plates");
$plates_sth->execute;
my $number_of_plates_processed = $plates_sth->fetchrow_array;
$plates_sth->finish;
$number_of_plates_processed || $page->error_page("No Overgo Plates found in physical database.");

# Print the page.
$page->header("The overgo plating process");
print page_title_html('About the overgo plating process');

print blue_section_html('The overgo plating process',<<EOH);
This page explains (briefly) the nature of the Overgo Plating experiments
conducted at SGN.  We also explain the terminology used on our
<a href="$overgo_stats_page">Overgo statistics page</a> in relation to the
various stages of processing we have conducted.
EOH

print blue_section_html('About the overgo plating process',<<EOH);
<p>Overgo plates have 96 wells, arranged in 8 rows by 12 columns.  DNA sequence
for one overgo probe (a sequenced marker from our maps collection) is assigned
to each well.  Wells are then <i>pooled</i> using the <a href="$soop_home">Soop</a>
program to produce 20 pools.  (Equivalent to 8 row-pools + 12 column-pools.)
</p>

<p>BAC plates are then run against these 20 pools for each plate to determine
which BACs match a given pool.  From this, the matches of BACs to probes
can be inferred.  A probe which successfully matches one or more BACs is
said to "anchor" them to the <a href="$physical_map_page">Genetic Map</a> and
is thus referred to as an "anchor point".
</p>


<p>To successfully anchor to a given probe, we require a BAC to:</p>

<ol>
<li>successfully match both its row-pool and column-pool, and</li>
<li>not match any other pools on that plate.</li>
</ol>

<p>In the case where BACs matched more than one row and one column on a given
plate, we classify them as being <i>ambiguous</i> BACs.  These are stored
separately and are not shown on our <a href="$physical_map_overview">physical 
map</a>.
</p>
<a name="fpc_contigging"></a>
EOH

print blue_section_html('The FPC process',<<EOH);
Meanwhile, working from the other end, our collaborators at the
<a href="$agi_page">Arizona Genomics Institute</a> are using <i>Fingerprint
Contigging (FPC)</i> techniques to assemble the BAC collection into contigs.
The BAC &lt;--&gt; map position matches generated by our overgo plating experiment
are used to inform the FPC process and improve the contigging conducted there.
EOH

print blue_section_html('In silico processing of BAC anchoring',<<EOH);
<p>After establishing the unambiguous BAC &lt;--&gt; probe associations, a further stage
of in silico processing was done to check for the <i>plausibility</i> of BAC
matches.  Our initial "plausible set" contains all unambiguously anchored
BACs.  We then remove BACs from this set which fail to meet certain criteria.
</p>

<p>In general, we expect a given BAC to only match up to a portion of one chromosome.
Thus, as a first step, we drop from our plausible set all BACs which are anchored
to markers on two or more chromosomes.  [<i>N.B.</i> - Once data from all plates are in,
it is our intention to conduct more complex analysis.  Specifically, if a BAC
has multiple anchor points within a tight range on one chromosome and only a
lone anchor point elsewhere then we may reasonably conclude that that one match
is the aberrancy.  However, at this stage the depth of data necessary to drive this 
analysis has not been accumulated.]
</p>

<p>Secondly, we expect BACs to be anchored largely to one portion of a chromosome,
rather than randomly anchored down its length.  Thus we require a plausible BAC
to have a "walk" of anchor points, all of which are within a reasonable distance
of one another.  Arbitrarily, we have chosen a distance of 5.0 cM as the maximum
distance between any adjacent anchor points for a given BAC.  BACs which are
in violation of this principle are dropped from our plausible set.
</p>

<p>Finally, we consider BAC contigs (as generated using FPC) to be plausible if:</p>

<ol>
<li>All of their member BACs which are anchored lie on the same chromosome, and</li>
<li>There is a "walk" of BACs down the length of the contig where no two anchor
points for the contig are more than 5.0 cM apart.</li>
</ol>
EOH

print blue_section_html('Summary of terms',<<EOH);
<p>This section gives a summary of the terms used on the 
<a href="$overgo_stats_page">Overgo Plating Statistics</a> page.  To
view the numbers for any of the following, please refer to that
page.</p>

<p>Terms are listed here in alphabetical order.</p>
</div>

<div class="indentedcontent">
<dl id="overgodefs">
<dt><a name="anchorpoint"></a>
Anchor Point
</dt>
<dd>
An overgo probe (qv) which has successfully been associated with one
or more BACs is said to anchor them to the map and is thus referred to as
an "anchor point".
</dd>

<dt><a name="bacs"></a>
BACs
</dt>
<dd>
BACs are <i>Bacterial Artificial Chromosomes</i>.  The same library of
BACs is
used for both the FPC contigging and the overgo plating process.  The
total number of BAC clones on the library plates is 129024, but not
all clones are valid and so the number of BACs reported is lower.
</dd>

<dt><a name="emptywells"></a>
Empty wells
</dt>
<dd>
Overgo plates (qv) typically have 96 wells.  In order to facilitate clear
matching between one probe in a given well, the dispersal of markers on
the plates has been <i>deconvoluted</i> so that no two markers sharing
a pool lie within 5.0 cM of one another on the <a 
href="$physical_map_page">Genetic Map</a>.  A consequence of this is that
it has sometimes been necessary to leave some wells empty on given plates
in order to ensure a thorough dispersal.  Thus, not every overgo plate
contains 96 probes.  Instead, some wells may remain empty.  The total
number of probes + empty wells for a given plate should add up to 96.
</dd>

<dt><a name="hittheplates"></a>
Hitting the plates
</dt>
<dd>
BACs which matched at least one overgo pool (corresponding to either a 
row or a column) on at least one of the overgo plates are said to have
"hit the plates", in that they have at least been found to in some way
match the sequences of the genetic markers known to SGN.
</dd>

<dt><a name="plates"></a>
Overgo Plates
</dt>
<dd>
Each overgo plate contains up to 96 probes, arranged in 8 rows and 12 columns.
To date, $number_of_plates_processed plates have been processed.
</dd>

<dt><a name="probes"></a>
Overgo Probes
</dt>
<dd>
Each <i>probe</i> refers to one sequenced marker found in the SGN tomato
maps collection.  Each marker is placed once on the overgo plates we
designed and counts as a probe.  If a probe successfully associates
BACs to the map then it is said to be an anchor point.
</dd>

<dt><a name="plausible"></a>
Plausible BAC locations
</dt>
<dd>
The overgo plating process may produce multiple associations to probes for a given
BAC.  If those probes all lie on the same chromosome, within a relatively 
well clustered region, then we say the BAC is <i>plausibly</i> anchored to
that chromosome by those anchor points.  The criterion for "well clustered"
is that no two consecutive anchor points are more than 5.0 cM apart.

An FPC contig whose plausibly anchored BACs all lie on the same chromosome
is said to be a <i>plausible contig</i>.
</dd>

<dt><a name="ambiguity"></a>
Unambiguous matching, a.k.a. plausible matching
</dt>
<dd>
<p>Ambiguity refers to the match between a given BAC and a given overgo plate.
If the BAC matches exactly one row-pool and one column-pool, thus specifying
a clear match to one probe on the plate, then it is taken to be unambiguously
anchored to that probe.  Otherwise, the set of possible probes that it could
be matched to is the set of ambiguous matches accorded to that BAC.</p>
</dd>

</dl>
EOH

print blue_section_html('Links',<<EOH);
<ul>
<li><a href="$overgo_stats_page">Progress of the Overgo Plating Project</a></li>
<li><a href="$agi_page">BAC Contigging by FPC at the Arizona Genomics Institute</a></li>
<li><a href="$physical_map_page">Physical map abstract page</a></li>
<li><a href="$physical_map_overview">Overview of the physical map</a></li>
<li><a href="$bac_search_page">Search SGN for BACs</a></li>
</ul>
EOH

$page->footer;
