use strict;
use CXGN::Page;
my $page=CXGN::Page->new('Labate_tomato_tech.html','html2pl converter');
$page->header('SNP Consortium Pre-proposal');
print<<END_HEREDOC;


    <h2>SNP Consortium Pre-proposal</h2>

    <p>Joanne Labate and Angela Baldo, USDA Geneva</p>

    <p>Cultivated tomato (<em>Lycopersicon esculentum</em> var. <em>esculentum</em>
    is known to have relatively low molecular genetic diversity.
    This limited genetic variation has restricted the use of molecular
    markers as tools for genetic studies or crop improvement. The
    USDA, ARS Plant Genetic Resources Unit (PGRU) in Geneva, NY
    currently conserves more than 5,000 accessions of cultivated
    tomato. We are developing DNA sequence-based molecular markers
    (Single Nucleotide Polymorphisms, known as SNPs) to
    characterize our collection. Polymorphic markers that we
    develop will also be useful to our stakeholders (e.g. tomato
    breeders) for saturating genetic maps and marker-assisted
    selection.</p>

    <p>Tomato is ideal for pioneering SNP prediction and confirmation because of the
    wealth of publicly available DNA sequence data. Baldo has
    designed and implemented a high-throughput Expressed Sequence
    Tag (EST) analysis system which takes a NCBI Unigene set as
    input, and produces high-quality annotation and consensus
    sequences of subclusters, virtual mapping of consensus
    sequences by matching to known markers, SNP predictions, and
    SSR discovery. An additional module designs and selects optimal
    PCR primers flanking the regions of interest. Using the more
    than 150,000 EST and cDNA sequences from over 15 cultivars
    comprising the 3,000+ member Unigene set for <em>L.
    esculentum</em> in Genbank, 2,527 SNPs in 764 Unigenes were predicted. In 2004 we tested 85
    independent amplicons from the 764 Unigenes for predicted SNPs
    by sequencing two or three cultivars per amplicon. We
    discovered 62 SNPs and 13 small insertion/deletion
    polymorphisms in 21 amplicons. For the 64 remaining amplicons,
    one primer pair did not amplify, thirty two showed no evidence
    of the predicted SNP, 20 appeared heterozygous or gave multiple
    PCR bands, and 11 gave insufficient data (poor quality
    sequence, fragment too large to sequence predicted SNP site, or
    data from only one cultivar). Based on the 53 amplicons that
    gave clear results thus far, this method discovered cultivated
    tomato SNPs with approximately 21-fold more efficiency compared
    to sequencing random genomic DNA (1 SNP per 300 nucleotides
    versus 1 SNP per 7 kb). We have submitted a manuscript for
    publication of the 21 polymorphic markers and are continuing development
    and testing of the remaining set of 64 of the 85 originally
    tested.</p>

    <p>We propose to
    develop and test the remaining 679 Unigenes with predicted SNPs
    in a collaborative effort with private companies such as
    Campbell's, Western Seed, Syngenta/Roger's Seed,
    Rijk Zwaan, Zeraim Gedera, and other interested parties. Based
    on our results we anticipate successfully developing at least
    25\% of the 764 Unigenes (191 loci) into polymorphic markers by
    sequencing two to three tomato cultivars with predicted SNPs
    among them. This will entail - obtaining the minimal set of
    cultivars needed based on NCBI's original EST data,
    growing the cultivars in PGRU's greenhouse and isolating
    DNA from leaf tissue, PCR primer design and PCR amplification
    for the remaining 679 Unigenes for two or three cultivars each,
    DNA sequencing, and analyses of the sequence data. The
    anticipated cost of carrying out this project in-house at PGRU
    is detailed in the attached budget pages. Products will be
    robust PCR and sequencing primers for genomic DNA sequences,
    identification of polymorphic sites among the assayed cultivars
    for those sequences, and annotation of sequences including
    predicted proteins and matches to known tomato markers. All
    results will be published in peer-reviewed journals and all
    collaborators will have access to preliminary and pre-published
    data.</p>

    <p>The proposed
    research is part of PGRU's "Conservation of
    Vegetable Crops" CRIS, Objective 2 Enhance the
    effectiveness of germplasm maintenance through the application
    of genomic sequencing and molecular marker techniques.
    It will serve as a model for discovery of enhanced value within crop
    germplasm when faced by limited molecular genetic diversity,
    ensuring the future viability of U.S. farms and a nutritious
    food supply.</p>
<br />
    <h3>Budget</h3>

    <table summary="" border="0" cellspacing="0" cellpadding="0" width="40\%">
      <tr>
        <td width="220"><strong><u>Item</u></strong></td>
        <td width="75"><strong><u>Year 1</u></strong></td>
        <td width="75"><strong><u>Year 2</u></strong></td>
        <td width="75"><strong><u>Total</u></strong></td>
      </tr>
<tr><td><p>&nbsp;</p></td></tr>
      <tr>
        <td><strong>Materials</strong></td>
        <td></td>
        <td></td>
        <td></td>
      </tr>
<tr><td><p>&nbsp;</p></td></tr>
      <tr>
        <td>PCR and DNA sequencing<br /> (\$4.95 per sample)</td>
        <td>\$12,997</td>
        <td>\$12,997</td>
        <td>\$25,994</td>
      </tr>
<tr><td><p>&nbsp;</p></td></tr>
      <tr>
        <td>Primers</td>
        <td>\$11,270</td>
        <td>\$1,691</td>
        <td>\$12,961</td>
      </tr>
<tr><td><p>&nbsp;</p></td></tr>
      <tr>
        <td>Instrument maintenance</td>
        <td>\$2,192</td>
        <td>\$2,192</td>
        <td>\$4,383</td>
      </tr>
<tr><td><p>&nbsp;</p></td></tr>
      <tr>
        <td><strong>Labor</strong></td>
        <td></td>
        <td></td>
        <td></td>
      </tr>
<tr><td><p>&nbsp;</p></td></tr>
      <tr>
        <td>Technician (GS-4)</td>
        <td></td>
        <td></td>
        <td></td>
      </tr>
<tr><td><p>&nbsp;</p></td></tr>
      <tr>
        <td>&nbsp;&nbsp;&nbsp;&nbsp;salary</td>
        <td>\$24,698</td>
        <td>\$26,288</td>
        <td>\$50,986</td>
      </tr>
<tr><td><p>&nbsp;</p></td></tr>
      <tr>
        <td>&nbsp;&nbsp;&nbsp;&nbsp;benefits</td>
        <td>\$7,409</td>
        <td>\$7,886</td>
        <td>\$15,296</td>
      </tr>
<tr><td><p>&nbsp;</p></td></tr>
      <tr>
        <td>3 Supervisory scientists<br /> (5\% each)</td>
        <td>\$16,778</td>
        <td>>\$17,615</td>
        <td>\$34,393</td>
      </tr>
<tr><td><p>&nbsp;</p></td></tr>
      <tr>
        <td><strong>Total materials and labor</strong></td>
        <td>\$75,344</td>
        <td>\$68,669</td>
        <td>\$144,013</td>	
      </tr>
<tr><td><p>&nbsp;</p></td></tr>
      <tr>
        <td><strong>Other Direct Costs</strong></td>
        <td></td>
        <td></td>
        <td></td>
      </tr>
<tr><td><p>&nbsp;</p></td></tr>
      <tr>
        <td>PGRU overhead IRC (0.1905)</td>
        <td>\$14,353</td>
        <td>\$13,081</td>
        <td>\$27,434</td>
      </tr>
<tr><td><p>&nbsp;</p></td></tr>
      <tr>
        <td><strong>Total Direct Costs</strong></td>
        <td>\$89,697</td>
        <td>\$81,750</td>
        <td>\$171,447</td>
      </tr>
<tr><td><p>&nbsp;</p></td></tr>
      <tr>
        <td><strong>Indirect Costs</strong></td>
        <td></td>
        <td></td>
        <td></td>
      </tr>
<tr><td><p>&nbsp;</p></td></tr>
      <tr>
        <td>ARS overhead (0.1111)</td>
        <td>\$9,965</td>
        <td>\$9,082</td>
        <td>\$19,048</td>
      </tr>
<tr><td><p>&nbsp;</p></td></tr>
      <tr>
        <td><strong>Total Direct and<br /> Indirect Costs</strong></td>
        <td><strong>\$99,662</strong></td>
        <td><strong>\$90,832</strong></td>
        <td><strong>\$190,495</strong></td>
      </tr>
    </table>

<br /><br /><br /><br />

<h4>Breakdown of Sequencing Costs</h4>

<table summary="" border="0" cellspacing="0" cellpadding="0" width="50\%">
    <tr>
    <td width="89\%">number of genes</td>
    <td align="right">700</td>
    </tr>
<tr><td><p>&nbsp;</p></td></tr>
    <tr>
    <td>number of cultivars</td>
    <td align="right">3</td>
    </tr>
<tr><td><p>&nbsp;</p></td></tr>
    <tr>
    <td>sequencing reactions (forward and reverse)</td>
    <td align="right">2</td>
    </tr>
<tr><td><p>&nbsp;</p></td></tr>
    <tr>
    <td>cost per sequence</td>
    <td>\$4.95</td>
    </tr>
<tr><td><p>&nbsp;</p></td></tr>
    <tr>
    <td>cost</td>
    <td>\$20,795</td>
    </tr>
<tr><td><p>&nbsp;</p></td></tr>
    <tr>
    <td>number of additional reactions (25\%) for redos</td>
    <td align="right">1050</td>
    </tr>
<tr><td><p>&nbsp;</p></td></tr>
    <tr>
    <td>cost for redos</td>
    <td>\$5,199</td>
    </tr>
<tr><td><p>&nbsp;</p></td></tr>
    <tr>
    <td>total sequencing cost</td>
    <td>\$25,994</td>
    </tr>
<tr><td><p>&nbsp;</p></td></tr>
    <tr>
    <td>total number of sequencing reactions</td>
    <td align="right">5250</td>
    </tr>
</table>

<br /><br />

<h4>Breakdown of Primer Costs</h4>

<table summary="" border="0" cellspacing="0" cellpadding="0" width="50\%">
    <tr>
    <td width="89\%">number of primer pairs</td>
    <td align="right">700</td>
    </tr>
<tr><td><p>&nbsp;</p></td></tr>
    <tr>
    <td>forward and reverse</td>
    <td align="right">2</td>
    </tr>
<tr><td><p>&nbsp;</p></td></tr>
    <tr>
    <td>size of primer</td>
    <td align="right">23</td>
    </tr>
<tr><td><p>&nbsp;</p></td></tr>
    <tr>
    <td>cost per base</td>
    <td>\$0.35</td>
    </tr>
<tr><td><p>&nbsp;</p></td></tr>
    <tr>
    <td>cost</td>
    <td>\$11,270</td>
    </tr>
<tr><td><p>&nbsp;</p></td></tr>
    <tr>
    <td>number of additional primers (15\%) added for redesigns</td>
    <td align="right">210</td>
    </tr>
<tr><td><p>&nbsp;</p></td></tr>
    <tr>
    <td>cost of redesigned primers</td>
    <td>\$1,691</td>
    </tr>
<tr><td><p>&nbsp;</p></td></tr>
    <tr>
    <td>total primer cost</td>
    <td>\$12,961</td>
    </tr>
<tr><td><p>&nbsp;</p></td></tr>
    <tr>
    <td>total number of primers</td>
    <td align="right">1610</td>
    </tr>
    </table>

<br /><br />

<h4>Estimated costs of PCR and sequencing per sample</h4>

<table summary="" border="0" cellspacing="0" cellpadding="0" width="70\%">
<tr>
<td width="25\%"></td>	
<td width="10\%"></td>
<td width="10\%"><strong>cost per unit</strong></td>
<td width="10\%"><strong>number used<br />per sample</strong></td>
<td width="10\%"><strong>cost per sample</strong></td>
</tr>
<tr><td><p>&nbsp;</p></td></tr>
<tr>
<td>10ul unplugged pipette tips</td>
<td></td>
<td>0.0173</td>
<td>1</td>
<td>0.02</td>
</tr>
<tr><td><p>&nbsp;</p></td></tr>
<tr>
<td>200ul unplugged pipette tips</td>
<td></td>
<td>0.014</td>
<td>2</td>
<td>0.03</td>
</tr>
<tr><td><p>&nbsp;</p></td></tr>
<tr>
<td>10ul plugged pipette tips</td>
<td></td>
<td>0.049</td>
<td>7</td>
<td>0.34</td>
</tr>
<tr><td><p>&nbsp;</p></td></tr>
<tr>
<td>200ul plugged pipette tips</td>
<td></td>
<td>0.063</td>
<td>2</td>
<td>0.13</td>
</tr>
<tr><td><p>&nbsp;</p></td></tr>
<tr>
<td>large orifice 200ul unplugged pipette tips</td>
<td></td>
<td>0.026</td>
<td>2</td>
<td>0.05</td>
</tr>
<tr><td><p>&nbsp;</p></td></tr>
<tr>
<td>large orifice 200ul plugged pipette tips</td>
<td></td>
<td>0.058</td>
<td>2</td>
<td>0.12</td>
</tr>
<tr><td><p>&nbsp;</p></td></tr>
<tr>
<td>50ul plugged pipette tips</td>
<td></td>
<td>0.05</td>
<td>2</td>
<td>0.10</td>
</tr>
<tr><td><p>&nbsp;</p></td></tr>
<tr>
<td>black 96-well clini plates</td>
<td></td>
<td>0.02</td>
<td>2</td>
<td>0.04
</td>
</tr>
<tr><td><p>&nbsp;</p></td></tr>
<tr>
<td>PCR plates</td>
<td></td>
<td>0.07</td>
<td>2</td>
<td>0.14</td>
</tr>
<tr><td><p>&nbsp;</p></td></tr>

<tr>
<td>1.5ml tubes</td>
<td></td>
<td>0.03</td>
<td>1</td>
<td>0.03</td>
</tr>
<tr><td><p>&nbsp;</p></td></tr>
<tr>
<td>Molecular Probes picogreen reagent<br />(1 includes forward and reverse)</td>
<td></td>
<td>0.14</td>
<td>0.5</td>
<td>0.07</td>
</tr>
<tr><td><p>&nbsp;</p></td></tr>
<tr>
<td>dNTPs</td>
<td></td>
<td>0.05</td>
<td>2.2</td>
<td>0.11</td>
</tr>
<tr><td><p>&nbsp;</p></td></tr>
<tr>
<td>Promega GoTaq PCR enzyme</td>
<td></td>
<td>0.22</td>
<td>2.2</td>
<td>0.48</td>
</tr>
<tr><td><p>&nbsp;</p></td></tr>
<tr>
<td>EdgeBio PCR clean-up<br />(1 includes forward and reverse)
</td>
<td>  
</td>
<td>1.03</td>
<td>0.5</td>
<td>0.52</td>
</tr>
<tr><td><p>&nbsp;</p></td></tr>
<tr>
<td>ABI BDT cycle sequencing enzyme</td>
<td>
</td>
<td>1.57</td>
<td>1</td>
<td>1.57</td>
</tr>
<tr><td><p>&nbsp;</p></td></tr>
<tr>
<td>EdgeBio sequencing rxn. clean-up</td>
<td></td>
<td>0.52</td>
<td>1</td>
<td>0.52</td>
</tr>   
<tr><td><p>&nbsp;</p></td></tr>
<tr>
<td>ABI sequencing polymer POP6</td>
<td></td>
<td>0.26</td>
<td>1</td>
<td>0.26</td>
</tr>
<tr><td><p>&nbsp;</p></td></tr>
<tr>
<td>ABI 3100 capillary array</td>
<td></td>
<td>0.43</td>
<td>1</td>
<td>0.20</td>
</tr>
<tr><td><p>&nbsp;</p></td></tr>
<tr>
<td></td>
<td></td>
<td></td>
<td align="right"><strong>total:</strong></td>
<td><strong>\$4.95</strong></td>
</tr>
</table>

<br /><br />

<h4>ABI 3100 DNA sequencer maintenance</h4>

<table summary="" border="0" cellspacing="0" cellpadding="0" width="50\%">
<tr>
<td width="16\%">lower polymer block</td>
<td width="1\%"></td>
<td width="1\%">\$2,200</td>
</tr>
<tr><td><p>&nbsp;</p></td></tr>
<tr>
<td>reserve polymer syringe</td>
<td></td>
<td>\$258</td>
</tr>
<tr><td><p>&nbsp;</p></td></tr>
<tr>
<td>array-fill syringe</td>
<td></td>
<td>\$150</td>
</tr>
<tr><td><p>&nbsp;</p></td></tr>
<tr>
<td>maintenance contract (10\% over 2 years)</td>
<td></td>
<td>\$1,775</td>
</tr>
<tr><td><p>&nbsp;</p></td></tr>
<tr>
<td></td>
<td align="right"><strong>total: </strong></td>
<td><strong>\$4,383</strong></td>
</tr>
</table>

<br /><br /><br /><br />

<h3>Budget narrative</h3>

<h4>Primers</h4>

    <p>Custom-synthesized oligos are purchased from
    Integrated DNA Technologies, Inc. PCR/sequencing oligos cost
    \$0.35/base at 25 nmole scale. Generally 4 pmol primer is used
    per PCR or sequencing reaction. We estimate 1,610 primers will
    be sufficient to PCR and sequence 700 loci taking into account
    needs for occasionally redesigning primers and replenishing
    primer stocks (15\% added). &nbsp; 23-mer x \$0.35 x 1,610 =
    \$12,961</p>

<h4>Laboratory costs</h4>

    <p>We routinely test all newly synthesized PCR primers across the 2
    or 3 genomic DNAs to be sequenced in small volume PCR
    reactions. After this initial optimization a larger volume (50
    ul) PCR reaction is performed, PCR products are cleaned using
    an EdgeBioSystems Quickstep kit, DNA is quantified in a
    picogreen assay, ABI BDT ver. 3.1 cycle sequencing reactions
    are performed, samples are cleaned on an EdgeBioSystems DTR
    plate, dried down, resuspended in formamide, and run on the ABI
    3100 capillary sequencer.</p>

    <p>Itemized costs per sample are in the budget
    table. Plastic disposables include pipette tips, 96-well
    plates, and eppendorf tubes. Disposable capillary arrays for
    the ABI 3100 cost \$695 each and last approximately 100 runs (1
    run = 16 samples). Costs of reagents including dNTPs, GoTaq and
    BDT enzymes, picogreen, POP6 polymer, and EdgeBio kits are
    estimated on a per sample basis. Cost per sequence = \$4.9513 x
    5,250 sequences = \$25,994</p>

<h4>Instrument maintenance</h4>

    <p>The throughput of our ABI 3100 DNA sequencer is two 96-well plates
    in 30 hours. For 5,250 reactions this is equivalent to 8 months
    of operation \@ 4 hours per day. We replace the lower polymer
    block, array-fill syringe, and reserve polymer syringe
    approximately every 6 months as routine maintenance. In 2004
    the ABI service maintenance contract cost \$8,750. We request
    \$2,192 per year to defray instrument maintenance
    costs.</p>

<h4>Labor</h4>

    <p>One full-time ARS Biological Science Technician (GS-4) is requested
    to carry out the majority of the labor for this project. Salary
    was estimated at entry level with 30\% added for benefits. There
    will be three supervisory scientists whose costs are estimated
    at 5\% of their salary and benefits each. Their respective roles
    can be described as follows - Vegetable Crops Curator Dr. Larry
    Robertson will be responsible for obtaining seed of required
    tomato cultivars, overseeing its planting and harvesting for
    tissue collection, and subsequent sample tracking and data
    curation. Larry has developed a database for PGRU for efficient
    tracking of samples from planting, through all laboratory
    assays, to final storage of molecular data. Bioinformaticist
    Dr. Angela Baldo will be responsible for all
    computational marker discovery: identification and downloading
    of public EST, cDNA, and genomic sequences, high-throughput
    in-house clustering, annotation, <em>in silico</em> mining for markers, and high-throughput
    primer design. Molecular Biologist Dr. Joanne Labate's
    primary responsibility on this project will be to train the
    technician in laboratory techniques and to oversee the
    collection and analyses of the molecular data, ensure quality
    control of laboratory-generated data, and dissemination of high
    quality data to collaborators.</p>

<h4>Overhead</h4>

    <p>Local overhead at PGRU is estimated as 16\% of total direct costs.
    National overhead for ARS is estimated as 10\% of total direct
    and indirect costs.</p>

<h4>Facilities and equipment</h4>

    <p>Physical facilities include three greenhouses with a total of 7500
    ft<sup>2</sup> for growing plants. The 1400 ft<sup>2</sup>
    laboratory is well-equipped to efficiently generate DNA
    sequence data. Major equipment includes an MJ Research Tetrad2 and two BioRad
    iCycler thermocyclers, one ABI 3100 and one ABI 310 Genetic
    Analyzer, a Tecan Genesis RSP 150 robotic liquid handling
    system, a plant tissue grinder (GenoGrinder), a
    refrigerated-centrifuge and a speed-vac that can hold 96-well
    plates, an Alpha Innotech FluorChem 8900 imaging and analysis
    system, and a 96-well plate reader for fluorescence and
    absorbance assays. We also have all necessary small equipment that we require such
    as microcentrifuges, incubators, shakers, autoclave, agarose
    gel rigs and power supplies, electronic multi-channel pipettes,
    -20&deg;C and -80&deg;C freezers, a freeze-drier,
    microcomputers, etc.</p>

    <p>We have at our
    disposal a Dell PowerEdge 6600 server with four 2.4GHz Intel
    side-bus processors, 16GB of RAM, and 876GB of storage in a
    RAID configuration, with an additional hot spare. This server
    currently runs Linux kernel 2.4, and shoulders the majority of
    the Unit's computational analyses. We have two secondary servers:
    a Dell 530 Workstation with a i686 single 1.7GHz processor
    server with 512MB memory, and 80GB of storage; Dell PowerEdge 4100 with
    200MHz processor with 256MB of memory and 22GB of storage, and
    a desktop workstation, all running Linux. Finally we have a
    Sony PCG-GRX700P notebook with i686 architecture, 2.20GHz,
    512MB RAM, and a 60GB hard drive, running a notebook-optimized
    version of the same Linux distribution. All five Linux machines are
    configured such that they can read and write to the shared
    Novell fileserver (Dell PowerEdge 2650, single processor
    1.8GHz, 1MB RAM, with 76GB of storage) and each other's hard
    drives. The rest
    of the computer network available for this project consists of
    approximately 14 Dell desktop PCs running Microsoft Windows
    connected by the Novell print and fileserver mentioned
    above.</p>

END_HEREDOC
$page->footer();